packages/adapter-utils/src/utils.ts

```diff
@@ -101,33 +125,102 @@ const fieldMapperGenerator = (
 
-export type TransformFuncSync = (args: TransformFuncArgs) => lowerDashTypes.NonPromise<Value> | undefined
+const recurseIntoValue = (
+  newVal: Value,
+  transformFunc: (value: Value, keyPathID?: ElemID, field?: Field) => Value,
+  strict: boolean,
+  allowEmpty: boolean,
+  keyPathID?: ElemID,
+  field?: Field,
+  fieldType?: TypeElement,
+): Value => {
+  if (newVal === undefined) {
+    return undefined
+  }
 
-type TransformValuesBaseArgs = {
-  values: Value
-  type: ObjectType | TypeMap | MapType | ListType
-  strict?: boolean
-  pathID?: ElemID
-  isTopLevel?: boolean
-  allowEmpty?: boolean
+  if (isReferenceExpression(newVal)) {
+    return newVal
+  }
+
+  if (field && isListType(fieldType)) {
+    const transformListInnerValue = (item: Value, index?: number): Value => (
+      // eslint-disable-next-line no-use-before-define
+      transformFunc(
+        item,
+        !_.isUndefined(index) ? keyPathID?.createNestedID(String(index)) : keyPathID,
+        new Field(
+          field.parent,
+          field.name,
+          fieldType.refInnerType,
+          field.annotations
+        ),
+      )
+    )
+    if (!_.isArray(newVal)) {
+      if (strict) {
+        log.debug(`Array value and isListType mis-match for field - ${field.name}. Got non-array for ListType.`)
+      }
+      return transformListInnerValue(newVal)
+    }
+    return newVal.map((v, i) => transformListInnerValue(v, i))
+  }
+  if (_.isArray(newVal)) {
+    // Even fields that are not defined as ListType can have array values
+    return newVal.map((item, index) => transformFunc(item, keyPathID?.createNestedID(String(index)), field))
+  }
+
+  if (isObjectType(fieldType) || isMapType(fieldType)) {
+    if (!_.isPlainObject(newVal)) {
+      if (strict) {
+        log.debug(`Value mis-match for field ${field?.name} - value is not an object`)
+      }
+      // _.isEmpty returns true for primitive values (boolean, number)
+      // but we do not want to omit those, we only want to omit empty
+      // objects, arrays and strings. we don't need to check for objects here
+      // because we cannot get here with an object
+      const valueIsEmpty = (Array.isArray(newVal) || _.isString(newVal)) && _.isEmpty(newVal)
+      return (valueIsEmpty && !allowEmpty) ? undefined : newVal
+    }
+    const fieldMapper = fieldMapperGenerator(fieldType, newVal)
+    return _.mapValues(
+      newVal,
+      (value, key) => transformFunc(value, keyPathID?.createNestedID(key), fieldMapper(key))
+    )
+  }
+  if (_.isPlainObject(newVal) && !strict) {
+    return _.mapValues(
+      newVal,
+      (value, key) => transformFunc(value, keyPathID?.createNestedID(key))
+    )
+  }
+  return newVal
 }
 
-type TransformValuesSyncArgs = TransformValuesBaseArgs & { transformFunc: TransformFuncSync }
-type TransformValuesArgs = TransformValuesBaseArgs & {
-  transformFunc: TransformFunc
-  elementsSource?: ReadOnlyElementsSource
+const depromise = async (x: Value): Promise<Value> => {
```

Comment:
shouldn't call things "x"
```suggestion
const depromise = async (value: Value): Promise<Value> => {
```